packages/protocol/src/abstracts/BaseRouter.sol

```diff
@@ -324,6 +309,195 @@ abstract contract BaseRouter is SystemAccessControl, IRouter {
+   * @param permit internal control number, either 1 for permitWithdraw, or 2 for permitBorrow
+   */
+  function _handlePermitAction(
+    bytes memory arg,
+    bytes32 actionArgsHash_,
+    uint256 permit
+  )
+    private
+    returns (address)
+  {
+    PermitArgs memory loaded;
+    {
+      (
+        loaded.vault,
+        loaded.owner,
+        loaded.receiver,
+        loaded.amount,
+        loaded.deadline,
+        loaded.v,
+        loaded.r,
+        loaded.s
+      ) = abi.decode(
+        arg, (IVaultPermissions, address, address, uint256, uint256, uint8, bytes32, bytes32)
+      );
+    }
+
+    _checkVaultInput(address(loaded.vault));
+
+    if (permit == 1) {
+      loaded.vault.permitWithdraw(
+        loaded.owner,
+        loaded.receiver,
+        loaded.amount,
+        loaded.deadline,
+        actionArgsHash_,
+        loaded.v,
+        loaded.r,
+        loaded.s
+      );
+    } else if (permit == 2) {
+      loaded.vault.permitBorrow(
+        loaded.owner,
+        loaded.receiver,
+        loaded.amount,
+        loaded.deadline,
+        actionArgsHash_,
+        loaded.v,
+        loaded.r,
+        loaded.s
+      );
+    }
+
+    return loaded.owner;
+  }
+
+  /**
+   * @dev Returns the `zeroPermitEncodedArgs` which is required to create
+   * the `actionArgsHash` used during permit signature
+   *
+   * @param vault that will execute action
+   * @param owner owner of the assets
+   * @param receiver of the assets after action
+   * @param amount of assets being permitted in action
+   */
+  function _getZeroPermitEncodedArgs(
+    IVaultPermissions vault,
+    address owner,
+    address receiver,
+    uint256 amount
+  )
+    private
+    pure
+    returns (bytes memory)
+  {
+    return abi.encode(vault, owner, receiver, amount, 0, 0, ZERO_BYTES32, ZERO_BYTES32);
+  }
+
+  /**
+   * @dev Returns the `actionsArgsHash` required in
+   * {VaultPermissions-permitWithdraw} or {VaultPermissions-permitBorrow}.
+   * Requirements:
+   * - Must replace arguments in IRouter.Action.PermitWithdraw for "zeroPermit".
+   * - Must replace arguments in IRouter.Action.PermitBorrow for "zeroPermit".
+   * - Must replace `beforeSlipped` amount in cross-chain txs that had slippage.
+   *
+   *
+   * @param actions being executed in this `_bundleInternal`
+   * @param args provided in `_bundleInternal`
+   * @param beforeSlipped amount passed by the origin cross-chain router operation
+   */
+  function _getActionArgsHash(
+    IRouter.Action[] memory actions,
+    bytes[] memory args,
+    uint256 beforeSlipped
+  )
+    private
+    pure
+    returns (bytes32)
+  {
+    uint256 len = actions.length;
+
+    /**
+     * @dev We intend to ONLY modify the new bytes array.
+     * "memory" in solidity persists within internal calls.
+     */
+    bytes[] memory modArgs = new bytes[](len);
+    for (uint256 i; i < len; i++) {
+      modArgs[i] = args[i];
+      if (
+        i == 0 && beforeSlipped != 0
```

Comment:
why the `i == 0` condition?

Reply:
Because you only replace slippage values in the first "value" transfer action in the destination chain. That means whatever comes from xReceive, either the token amount you want to deposit or to payback.